include/alpaka/queue/QueueHipRtBlocking.hpp

```diff
@@ -272,6 +274,9 @@ namespace alpaka
                     std::thread t(
                         [pCallbackSynchronizationData, task, &queue](){
 
+                            // thread-safe task execution and callee decrementing
+                            std::lock_guard<std::mutex> guard(queue.m_spQueueImpl->m_mutex);
```

Comment:
Why can't this lock be directly in front of the `queue.m_spQueueImpl->m_callees -= 1;` line?

Reply:
It actually can, as the lock is not relevant for the local object `pCallbackSynchronizationData`.

Reply:
Reverting last comment: I have to lock the whole region, otherwise the task() execution is not thread-safe. It is actually mentioned in the comment. The queue tests will fail otherwise (queueShouldNotExecuteTasksInParallel).

Reply:
If this is really necessary, then it is either a bug in the HIP implementation or a documentation error see
https://rocm-documentation.readthedocs.io/en/latest/ROCm_API_References/HIP_API/Stream-Management.html#hipstreamaddcallback: `The callback will block later work in the stream until it is finished.`

Reply:
I linked your comment in #794 so we can work on that issue separate. Since @tdd11235813 will have 3 week vocation next week I like to merge this PR to be able to use ROCm.

Reply:
thanks, I can squash some commits (Queue***Impl thing), if you want.

Regarding the bug, had no time to report with a minimal example, but it looks clear to me, that the stream is not blocked by the invoked callback execution, which is different to the cuda part, so I do not think, it is intended by HIP. I will drop a simple issue on the HIP side, so it is at least documented.